src/backend/core/grafana_events.py

```
@@ -0,0 +1,254 @@
+# pass so the dict does not grow for the process lifetime.
+
+_guardrail_attempts: dict = {}
+_guardrail_lock = threading.Lock()
+
+
+def _current_workflow_id() -> str:
+    """Bound workflow_id, or "" outside a workflow. Never raises."""
+    try:
+        import structlog.contextvars
+        return str(structlog.contextvars.get_contextvars().get("workflow_id", ""))
+    except Exception:
```

Ruff (0.15.21)

[warning] Do not catch blind exception: `Exception`

(BLE001)